modules/connectivity/ParticipantConnectionStatus.js

```diff
@@ -380,18 +424,53 @@ export default class ParticipantConnectionStatus {
             isConnActiveByJvb = true;
         }
 
-        const isConnectionActive
-            = isConnActiveByJvb
-                && (isVideoMuted || (isInLastN && !isVideoTrackFrozen));
+        let newState = ParticipantConnectionStatus.INACTIVE;
+
+        if (isConnActiveByJvb) {
+            if (isInLastN) {
+                if (isVideoTrackFrozen) {
+                    // if participant was inactive, and is in lastN,
+                    // it is restoring
+                    const isInactive = participant.getConnectionStatus()
+                        === ParticipantConnectionStatus.INACTIVE;
+
+                    newState = isInactive
```

Comment:
The definition of RESTORING is being inactive and entering lastN, so I do not see a way of calculating this without checking previous state. 
I agree with your point about calling it twice and that it can stay in RESTORING in certain cases, will fix those.
I will check the p2p, this was written before p2p was merged. I was assuming that p2p will work as and when lastN is disabled, all participants will be in lastN, is this correct?